### PR TITLE
Fix onDiscard leaks in fused operators with queue

### PR DESCRIPTION
**Pull request by OlegDokuka**
Follow up on on the series of `onDiscard` leaks fixes for specific fusion case

Signed-off-by: Oleh Dokuka <shadowgun@i.ua>

### CODE REVIEW

**Comment by OlegDokuka** [comment]
@simonbasle few cases are failing still, investigation what I did broker. However, you may get a general idea looking into the changes. I foresee that as a common pattern for all following changes in fuseable operators

File: reactor-core/src/test/java/reactor/core/publisher/OnDiscardShouldNotLeakTest.java [hunk 1]
```diff
@@ -45,24 +45,50 @@
...
-			DiscardScenario.fluxSource("unicastProcessor", 1, f -> f.subscribeWith(UnicastProcessor.create()))
-	};
-
-	private static boolean[][] CONDITIONAL_AND_FUSED = new boolean[][] {
-			{ false, false },
-			{ true, false },
-			{ false, true },
-			{ true, true }
-	};
+    //add DiscardScenarios here to test more operators
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/OnDiscardShouldNotLeakTest.java:48]: formatting: there shouldn't be spaces in this file if you use the official Reactor intellij format

Comment (OlegDokuka) [reply reactor-core/src/test/java/reactor/core/publisher/OnDiscardShouldNotLeakTest.java:48]: 🙃

File: reactor-core/src/main/java/reactor/core/publisher/FluxOnBackpressureBuffer.java [hunk 2]
```diff
@@ -349,7 +364,29 @@ public boolean isEmpty() {
...
+			if (DISCARD_GUARD.getAndIncrement(this) != 0) {
+				return;
+			}
+
+			int missed = 1;
+
+			for (;;) {
+				Operators.onDiscardQueueWithClear(queue, ctx, null);
+
+				int w = discardGuard;
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/FluxOnBackpressureBuffer.java:379]: nitpick: can you rename the variable to at least stick to the convention that local copies in drain loops are named after the initial of their corresponding field? it. `w` becomes `dg`

File: reactor-core/src/main/java/reactor/core/publisher/FluxOnBackpressureBuffer.java [hunk 3]
```diff
@@ -349,7 +364,29 @@ public boolean isEmpty() {
 
 		@Override
 		public void clear() {
-			Operators.onDiscardQueueWithClear(queue, ctx, null);
+			// use guard on the queue instance as the best way to ensure there is no racing on draining
+			// the call to this method must be done only during the ASYNC fusion so all the callers will be waiting
+			// this should not ber performance costly with the assumption the cancel is rare operation
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/core/publisher/FluxOnBackpressureBuffer.java:369]: typo `not ber` -> `not be`

File: reactor-core/src/test/java/reactor/core/publisher/OnDiscardShouldNotLeakTest.java [hunk 4]
```diff
@@ -45,24 +45,50 @@
...
-			DiscardScenario.fluxSource("unicastProcessor", 1, f -> f.subscribeWith(UnicastProcessor.create()))
-	};
-
-	private static boolean[][] CONDITIONAL_AND_FUSED = new boolean[][] {
-			{ false, false },
-			{ true, false },
-			{ false, true },
-			{ true, true }
-	};
+    // add DiscardScenarios here to test more operators
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/OnDiscardShouldNotLeakTest.java:48]: still whitespace formatting issue

File: reactor-core/src/main/java/reactor/core/publisher/FluxOnBackpressureBuffer.java (hunk 3 above)
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/FluxOnBackpressureBuffer.java:369]: applies to all occurrences below as well

File: reactor-core/src/main/java/reactor/core/publisher/FluxOnBackpressureBuffer.java (hunk 2 above)
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/publisher/FluxOnBackpressureBuffer.java:379]: applies to all occurrences below as well

File: reactor-core/src/test/java/reactor/core/publisher/OnDiscardShouldNotLeakTest.java (hunk 4 above)
Comment (OlegDokuka) [reply reactor-core/src/test/java/reactor/core/publisher/OnDiscardShouldNotLeakTest.java:48]: fixed

File: reactor-core/src/main/java/reactor/core/publisher/FluxOnBackpressureBuffer.java (hunk 3 above)
Comment (OlegDokuka) [reply reactor-core/src/main/java/reactor/core/publisher/FluxOnBackpressureBuffer.java:369]: fixed

File: reactor-core/src/main/java/reactor/core/publisher/FluxOnBackpressureBuffer.java (hunk 2 above)
Comment (OlegDokuka) [reply reactor-core/src/main/java/reactor/core/publisher/FluxOnBackpressureBuffer.java]: fixed

File: reactor-core/src/test/java/reactor/core/publisher/OnDiscardShouldNotLeakTest.java [hunk 5]
```diff
@@ -45,7 +45,7 @@
 @RunWith(Parameterized.class)
 public class OnDiscardShouldNotLeakTest {
 
-    // add DiscardScenarios here to test more operators
+	// add DiscardScenarios here to test more operators
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/OnDiscardShouldNotLeakTest.java:48]: haha nice try, I meant on the whole file 😛

Comment (OlegDokuka) [reply reactor-core/src/test/java/reactor/core/publisher/OnDiscardShouldNotLeakTest.java:48]: formated the whole file :D

**Pull request merged**